### PR TITLE
ENH: improvements to optimize.fmin_bfgs (PR for old issue / commit)

### PR DESCRIPTION
**Pull request by WillTirone**
Submitting a PR for community review / to bring some attention to the issue.

#### Reference issue 
Closes #1261 (old Trac issue)

#### What does this implement/fix?
Per original post: 
* the second linesearch can't return an alpha_k != None because it gets 'wrong' old_fval,old_old_fval. These get changed on the return from the first linesearch. These return values should be stored temporarily and only be used of the first linesearch returns an useable alpha_k
* a more stable way to compute sk

#### Additional information
[I  wrote a summary here](https://github.com/scipy/scipy/issues/1261#issuecomment-1146675187) that might be helpful to review what we need to do for the change.

### CODE REVIEW

**Comment by WillTirone** [comment]
@mdhaber I wasn't sure if it was as simple as literally just opening a PR from Ralf's original commit, but that's what I did above. Let me know if you had anything else in mind. Thanks!

**Review by mdhaber — COMMENTED** [review]
Well, it's little more involved because the filename of the file with this function has changed from `optimize.py` to `_optimize.py`. The merge conflicts need to be resolved. 

The changes themselves look reasonable. But I should re-apply these comments after the merge conflicts are resolved.  My suggestion is to locally resolve the comments by:
* `git remote add rgommers git@github.com:rgommers/scipy.git` to add Ralf's fork as a remote
* `git checkout rgommers/ticket-734-fmin_bfgs` to check out Ralph's branch (in a "detached HEAD state, I think". Doesn't really matter either way.)
* Assuming you aren't able to push to Ralf's branch, `git checkout -b gh1261` to create your own copy of the branch.
* `git fetch upstream` and `git merge upstream/main` to merge main. There will be conflicts.
* `git status`. Confirm that the only file with conflicts is `optimize.py`.
* `git checkout upstream/main scipy/optimize/optimize.py` to undo the changes to `optimize.py`.
* Manually add the changes to `_optimize.py` as closely as possible to the original changes in `optimize.py`.
* `git commit`

Then make this PR from your own branch. I think you have to create a new PR from your gh1261 branch, really.

From your summary of the original issue:
image

Sure, a damped update might be better, but this is an improvement as-is. IMO it is better to merge an improvement now than to wait for an better solution. 

We still need a test for `xtol`. I think we can construct one by setting all the other tolerances very loose on a simple problem and checking that it terminates due to `xtol` instead of the others.

**Comment by WillTirone** [comment]
Ahh okay, that all makes much more sense, let me attempt that over the next week or so and I will let you know what I find. Closing this PR and I'll make a new one as suggested above.

File: scipy/optimize/optimize.py [hunk 1]
```diff
@@ -509,18 +512,25 @@ def fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-5, norm=Inf,
         gnorm = vecnorm(gfk,ord=norm)
         if (gnorm <= gtol):
             break
+        # stopping criteria following http://www2.imm.dtu.dk/~hbn/Software/#GO
+        #  See Chapter 5 in  P.E. Frandsen, K. Jonasson, H.B. Nielsen,
+        #  O. Tingleff: "Unconstrained Optimization", IMM, DTU.  1999.
+        #  These notes are available here:
+        #   http://www2.imm.dtu.dk/documents/ftp/publlec.html
```
[the remark this answers is not in the document]
Comment (WillTirone) [reply scipy/optimize/optimize.py:519]: Can't seem to find anything more specific after a cursory look through the book.

**Pull request closed without merging**